xilem/src/view/flex.rs

```diff
@@ -139,61 +141,255 @@ where
+        let mut flex = self.element.borrow_mut();
+        match element {
+            FlexElement::Child(child, params) => {
+                flex.insert_flex_child_pod(self.idx, child.inner, params);
+            }
+            FlexElement::FixedSpacer(len) => flex.insert_spacer(self.idx, len),
+            FlexElement::FlexSpacer(len) => flex.insert_flex_spacer(self.idx, len),
+        };
+        self.idx += 1;
     }
-    fn with_scratch<R>(&mut self, f: impl FnOnce(&mut AppendVec<Pod<Box<dyn Widget>>>) -> R) -> R {
+
+    fn with_scratch<R>(&mut self, f: impl FnOnce(&mut AppendVec<FlexElement>) -> R) -> R {
         let ret = f(&mut self.scratch);
+        let mut flex = self.element.borrow_mut();
         for element in self.scratch.drain() {
-            self.element.insert_child_pod(self.ix, element.inner);
-            self.element.insert_default_spacer(self.ix + 1);
-            self.ix += 2;
+            match element {
+                FlexElement::Child(child, params) => {
+                    flex.insert_flex_child_pod(self.idx, child.inner, params);
+                }
+                FlexElement::FixedSpacer(len) => flex.insert_spacer(self.idx, len),
+                FlexElement::FlexSpacer(len) => flex.insert_flex_spacer(self.idx, len),
+            };
+            self.idx += 1;
         }
         ret
     }
 
-    fn mutate<R>(
-        &mut self,
-        f: impl FnOnce(<Pod<Box<dyn Widget>> as xilem_core::ViewElement>::Mut<'_>) -> R,
-    ) -> R {
-        let child = self
-            .element
-            .child_mut(self.ix)
-            .expect("ElementSplice::mutate won't overflow");
+    fn mutate<R>(&mut self, f: impl FnOnce(Mut<'_, FlexElement>) -> R) -> R {
+        let child = FlexElementMut {
+            parent: self.element.clone(),
+            idx: self.idx,
+        };
         let ret = f(child);
-        // Skip past the implicit spacer as well as this child
-        self.ix += 2;
+        self.idx += 1;
         ret
     }
 
-    fn delete<R>(
-        &mut self,
-        f: impl FnOnce(<Pod<Box<dyn Widget>> as xilem_core::ViewElement>::Mut<'_>) -> R,
-    ) -> R {
-        let child = self
-            .element
-            .child_mut(self.ix)
-            .expect("ElementSplice::mutate won't overflow");
+    fn delete<R>(&mut self, f: impl FnOnce(Mut<'_, FlexElement>) -> R) -> R {
+        let child = FlexElementMut {
+            parent: self.element.clone(),
+            idx: self.idx,
+        };
         let ret = f(child);
-        self.element.remove_child(self.ix);
-        // Also remove the implicit spacer
-        // TODO: Make the spacers be explicit?
-        self.element.remove_child(self.ix);
-
+        let mut flex = self.element.borrow_mut();
+        flex.remove_child(self.idx);
         ret
     }
 
     fn skip(&mut self, n: usize) {
-        self.ix += n * 2;
+        self.idx += n;
+    }
+}
+
+pub struct FlexItem<V, State, Action> {
```

Comment:
Because of parameterization issues, I needed to add these generic parameters.

I guess we need to do this a lot more for other views as well, to seamlessly support stuff like `Adapt` or in general inference for these variables, when it's not explicit somewhere...

Reply:
Yeah, this is a choice we have to make, and I think adding them to the View type is the right one, to keep the type inference failures local.